client/my-sites/plans/current-plan/jetpack-product-install/index.js

```diff
@@ -0,0 +1,239 @@
+ * We attempt to recover from these errors by retrying status requests.
+ */
+const RECOVERABLE_ERROR_STATES = [ 'vaultpress_error' ];
+/**
+ * The plugins this product installer installs, activates and configures.
+ */
+const PLUGINS = [ 'akismet', 'vaultpress' ];
+/**
+ * Maximum number of attempts to refetch installation status in the event of a recoverable error.
+ */
+const MAX_RETRIES = 3;
+
+export class JetpackProductInstall extends Component {
+	state = {
+		startedInstallation: false,
+	};
+
+	retries = 0;
+
+	componentDidMount() {
+		this.requestInstallationStatus();
+		this.maybeStartInstall();
+	}
+
+	componentDidUpdate() {
+		this.maybeStartInstall();
+	}
+
+	/**
+	 * Start the plugin installation if all conditions are matched:
+	 * - Installation hasn't started yet.
+	 * - We don't have any non-recoverable installation errors.
+	 * - We have already fetched the plugin installation status.
+	 * - Installation has not finished.
+	 * - We already have the plugin keys.
+	 */
+	maybeStartInstall() {
+		const { pluginKeys, progressComplete, siteId } = this.props;
+
+		// We're already installing
+		if ( this.state.startedInstallation ) {
+			return;
+		}
+
+		// We have non-recoverable installation errors
+		if ( this.installationHasErrors() && ! this.installationHasRecoverableErrors() ) {
+			return;
+		}
+
+		// We need to wait for installation status first
+		if ( progressComplete === null ) {
+			return;
+		}
+
+		// We should skip install if everything is already installed and configured
+		if ( progressComplete === 100 ) {
+			return;
+		}
+
+		// Installation can be started only if we have the Akismet and VaultPress keys
+		if ( pluginKeys && pluginKeys.akismet && pluginKeys.vaultpress ) {
+			this.setState( {
+				startedInstallation: true,
+			} );
+
+			this.props.startJetpackProductInstall( siteId, pluginKeys.akismet, pluginKeys.vaultpress );
+		}
+	}
+
+	/**
+	 * Used to determine if at least one plugin is in at least one of the provided plugin states.
+	 *
+	 * @param {Array} pluginStates States to check against.
+	 * @return {Boolean} True if at least one plugin is in at least one of the given states, false otherwise.
+	 */
+	arePluginsInState( pluginStates ) {
+		const { status } = this.props;
+
+		if ( ! status ) {
+			return false;
+		}
+
+		return some( PLUGINS, pluginSlug => {
+			if ( pluginStates.indexOf( status[ pluginSlug + '_status' ] ) >= 0 ) {
+				return true;
+			}
+			return false;
+		} );
+	}
+
+	/**
+	 * Used to determine if at least one plugin is in an error state.
+	 * Potential errors we consider here could be recoverable or not.
+	 * What we don't consider errors are the `NON_ERROR_STATES` above.
+	 *
+	 * @return {Boolean} Whether there are currently any installation errors.
+	 */
+	installationHasErrors() {
+		if ( this.installationHasRecoverableErrors() ) {
+			return true;
+		}
+
+		return ! this.arePluginsInState( NON_ERROR_STATES );
```

Comment:
Would it make sense to move this line to a new separate method `installationHasIrrecoverableErrors`, and use that on [L87](https://github.com/Automattic/wp-calypso/pull/31684/files#diff-0fb0ce86f509d38d8680f5a4c65dc4fdR87)?

Reply:
Hm, while this looks like it makes sense in the first place, it might make the logic more confusing (and actually to contain wrong assumptions), because then `installationHasIrrecoverableErrors` would know that we have an error, but wouldn't know it's irrecoverable without actually checking `installationHasRecoverableErrors`. So it would end up being a twin of `installationHasErrors`, so I prefer leaving as-is, if you don't mind.